backend-rust/src/indexer.rs

```diff
@@ -1340,3 +1380,134 @@ impl PreparedBakerEvent {
         Ok(())
     }
 }
+
+struct PreparedModuleDeployed {
+    block_height: i64,
+    deployment_transaction_index: i64,
+    module_reference: String,
+    schema: Option<String>,
+}
+
+impl PreparedModuleDeployed {
+    async fn prepare(
+        node_client: &mut v2::Client,
+        data: &BlockData,
+        block_item: &BlockItemSummary,
+        module_reference: sdk_types::hashes::ModuleReference,
+    ) -> anyhow::Result<Self> {
+        let block_height = data.finalized_block_info.height;
+        let deployment_transaction_index = block_item.index.index.try_into()?;
+
+        let wasm_module = node_client
+            .get_module_source(&module_reference, BlockIdentifier::AbsoluteHeight(block_height))
+            .await?
+            .response;
+        let schema = match wasm_module.version {
+            WasmVersion::V0 => get_embedded_schema_v0(wasm_module.source.as_ref()),
+            WasmVersion::V1 => get_embedded_schema_v1(wasm_module.source.as_ref()),
+        }
+        .ok();
+
+        let schema = schema.as_ref().map(|s| s.to_string());
+
+        Ok(Self {
+            block_height: i64::try_from(block_height.height)?,
+            deployment_transaction_index,
+            module_reference: module_reference.into(),
+            schema,
+        })
+    }
+
+    async fn save(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<()> {
+        sqlx::query!(
+            r#"
+INSERT INTO smart_contract_modules (
+    index,
+    module_reference,
+    deployment_block_height,
+    deployment_transaction_index,
+    schema
+) VALUES (
+  (SELECT COALESCE(MAX(index) + 1, 0) FROM smart_contract_modules),
```

Comment:
Okay this makes me think that the index should use an auto-incrementing serial field, then you don't need to manually set this (and using this method is dangerous inside transactions). See the serial types here: https://www.postgresql.org/docs/current/datatype-numeric.html

Reply:
The advantage with this approach is that it will ensure the index does not skip any numbers, and it should be safe since we don't insert these in parallel transactions.
Since smart contracts are using the `module_reference`, we don't really need this field anyway, so it should probably just be removed.
FYI [Identity columns](https://www.postgresql.org/docs/current/ddl-identity-columns.html) seems to be better than `Serial`.

Reply:
It just seems kind of brittle; how do we ensure that parallel insertions are not performed, also in the future? I guess parallel insertions would just fail since they'd try to insert the same index. Are we okay with that?

We could consider taking an exclusive lock on the table whenever we insert so that at least any parallel insertions would not fail, but would just be forced to be sequential. But I guess we can also forget to take such a lock.

Reply:
Supporting parallel insertions would introduce a lot of complexity everywhere in the indexer, and I doubt it would be worth it in terms of performance.
We already ensure the inserts are done sequentially, so it should only happen if you run several instances of the indexer, which would cause other problems as well.
This we could improve by introducing some [Advisory lock](https://www.postgresql.org/docs/current/explicit-locking.html#ADVISORY-LOCKS), but not as part of this PR and probably not for the first iteration of this backend.

Reply:
It might make sense to document this limitation/design choice in the readme.

Reply:
Like the discussion in the thread.

@Victor-N-Suadicani 
Feel free to document the design decisions in the Readme. For now, I want to contain the scope of this PR to the issues it is addressing and closing this thread.